### PR TITLE
Transparent attribute to specify that representation is the same as its only field

### LINKED ISSUE #1054: Attribute to specify that type representation is the same as its only field
By analogy with [`#[repr(transparent)]`](https://github.com/rust-lang/rfcs/blob/master/text/1758-repr-transparent.md).

```rust
// This should serialize and deserialize directly as String, rather than as newtype.
#[derive(Serialize, Deserialize)]
#[serde(transparent)]
struct Transparent(String);

// Generated code.
impl Serialize for Transparent {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
        where S: Serializer
    {
        self.0.serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for Transparent {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
        where D: Deserializer<'de>
    {
        Deserialize::deserialize(deserializer).map(Transparent)
    }
}
```

Should also work for structs with one field, and structs and tuple structs with all but one skipped field.

@bluss

### PR DESCRIPTION
**Pull request by dtolnay**
Fixes #1054.

### CODE REVIEW

File: serde_derive/src/internals/check.rs [hunk 1]
```diff
@@ -278,6 +279,31 @@ fn check_adjacent_tag_conflict(cx: &Ctxt, cont: &Container) {
     }
 }
 
+/// Enums and unit structs cannot be transparent.
```
Comment (oli-obk) [inline serde_derive/src/internals/check.rs]: Technically single variant enums are just structs

File: serde_derive/src/internals/attr.rs [hunk 2]
```diff
@@ -777,7 +789,6 @@ pub enum Default {
 }
 
 impl Default {
-    #[cfg(feature = "deserialize_in_place")]
```
Comment (oli-obk) [inline serde_derive/src/internals/attr.rs]: Is this change intended?

File: serde_derive/src/ser.rs [hunk 3]
```diff
@@ -186,6 +188,52 @@ fn serialize_body(cont: &Container, params: &Parameters) -> Fragment {
...
+        }
+        if transparent_field.is_some() {
+            panic!("Ambiguous transparent field");
+        }
+        transparent_field = Some(field);
+    }
+
+    match transparent_field {
+        Some(transparent_field) => transparent_field,
+        None => panic!("No field can be transparent"),
```
Comment (oli-obk) [inline serde_derive/src/ser.rs]: That message feels suboptimal. I know why it's worded this way because I'm reading the PR, but still found it confusing

File: serde_derive/src/internals/check.rs (hunk 1 above)
Comment (dtolnay) [reply serde_derive/src/internals/check.rs]: True but let's leave that until someone asks for it.

File: serde_derive/src/internals/attr.rs (hunk 2 above)
Comment (dtolnay) [reply serde_derive/src/internals/attr.rs]: I believe so, deciding whether a field during `derive(Deserialize)` is transparent now checks `field.attrs.default().is_none()`.

File: serde_derive/src/ser.rs (hunk 3 above)
Comment (dtolnay) [reply serde_derive/src/ser.rs]: Thanks, I improved the message and added some compile-fail tests.
